File: Task Tracker C#/contest/contestNetworking/ContestProtoProxy.cs
using Com.Example.Networking;
using contestService;
using Google.Protobuf;
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading.Tasks;


namespace contestNetworking
{
    public class ContestProtoProxy : IContestService
    {
        private string host;
        private int port;

        private IContestObserver client;

        private NetworkStream stream;

        private TcpClient connection;

        private Queue<ContestResponse> responses;
        private volatile bool finished;
        private EventWaitHandle _waitHandle;

        public ContestProtoProxy(string host, int port)
        {
            this.host = host;
            this.port = port;
            responses = new Queue<ContestResponse>();
        }

        private bool IsUpdate(ContestResponse response)
        {
            return response.Rep == ContestResponse.Types.Reply.Update;
        }

        private void HandleUpdate(ContestResponse response)
        {
            Console.WriteLine("Updating...");
            try
            {
                client.updateTaskList();
            }
            catch (ContestException cx)
            {
                Console.WriteLine(cx.StackTrace);
            }
        }

        private void StartReader()
        {
            Thread tw = new Thread(Run);
            tw.Start();
        }

        private void InitializeConnection()
        {
            try
            {
                connection = new TcpClient(host, port);
                stream = connection.GetStream();
                finished = false;
                _waitHandle = new AutoResetEvent(false);
                StartReader();
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        private void CloseConnection()
        {
            finished = true;
            try
            {
                stream.Close();
                connection.Close();
                _waitHandle.Close();
                client = null;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        private void SendRequest(ContestRequest request)
        {
            try
            {
                request.WriteDelimitedTo(stream);
                stream.Flush();
            }
            catch (Exception e)
            {
                throw new ContestException("Error sending object " + e);
            }
        }

        private ContestResponse ReadResponse()
        {
            ContestResponse response = null;
            try
            {
                _waitHandle.WaitOne();
                lock (responses)
                {
                    response = responses.Dequeue();
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
            return response;
        }

        public virtual void Run()
        {
            while (!finished)
            {
                try
                {
                    ContestResponse response = ContestResponse.Parser.ParseDelimitedFrom(stream);
                    Console.WriteLine("Response received " + response);
                    if (IsUpdate(response))
                    {
                        HandleUpdate(response);
                    }
                    else
                    {
                        lock (responses)
                        {
                            responses.Enqueue(response);
                        }
                        _waitHandle.Set();
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine("Reading error " + e);
                }
            }
        }

        public virtual void Login(contestDomain.User user, IContestObserver obs)
        {
            InitializeConnection();
            Console.WriteLine("Login proxy");
            SendRequest(ProtoUtils.CreateLoginRequest(user));
            ContestResponse response = ReadResponse();
            Console.WriteLine("Login proxy after response");
            if (response.Rep == ContestResponse.Types.Reply.Ok)
            {
                this.client = obs;
                return;
            }
            if (response.Rep == ContestResponse.Types.Reply.Error)
            {
                string err = ProtoUtils.GetError(response);
                CloseConnection();
                throw new ContestException(err);
            }
        }

        public virtual void Logout(contestDomain.User user, IContestObserver observer)
        {
            SendRequest(ProtoUtils.CreateLogoutRequest(user));
            ContestResponse response = ReadResponse();
            CloseConnection();
            if (response.Rep == ContestResponse.Types.Reply.Error)
            {
                string err = ProtoUtils.GetError(response);
                throw new ContestException(err);
            }
        }

        private List<contestDomain.dto.TaskDTO> Model2ProtoMapper(List<TaskDTO> tasks)
        {
            List<contestDomain.dto.TaskDTO> res = new List<contestDomain.dto.TaskDTO>();
            foreach (var tk in tasks)
            {
                contestDomain.dto.TaskDTO taskDTO = new contestDomain.dto.TaskDTO(tk.TaskID, (contestDomain.enums.Type)Enum.Parse(typeof(contestDomain.enums.Type), tk.Type.ToString().ToUpper()), (contestDomain.enums.AgeGroup)Enum.Parse(typeof(contestDomain.enums.AgeGroup), tk.Age.ToString().ToUpper()), tk.Enrolled);
                res.Add(taskDTO);
            }
            return res;
        }

        public virtual List<contestDomain.dto.TaskDTO> GetTasksEnrolled()
        {
            Console.WriteLine("Requesting all the tasks...");
            Request request = new GetTasksEnrolled();
            SendRequest(ProtoUtils.CreateTasksEnrolledRequest());
            ContestResponse response = ReadResponse();
            Console.WriteLine("Finished reading the get all response...");
            if(response.Rep == ContestResponse.Types.Reply.GetAllTasksExperimental)
            {
                return Model2ProtoMapper(ProtoUtils.GetEnrolledTasks(response));
            }
            if (response.Rep == ContestResponse.Types.Reply.Error)
            {
                string err = ProtoUtils.GetError(response);
                throw new ContestException(err);
            }
            return null;
        }

        private List<contestDomain.Participant> Model2ProtoParticipantMapper(List<ParticipantDTO> parts)
        {
            List<contestDomain.Participant> res = new List<contestDomain.Participant>();
            foreach (var tk in parts)
            {
                contestDomain.Participant p = new contestDomain.Participant(new Random().Next(500, 1488), tk.Name, tk.Age);
                res.Add(p);
            }
            return res;
        }

        public virtual List<contestDomain.Participant> GetFilteredParticipants(contestDomain.enums.Type type, contestDomain.enums.AgeGroup age)
        {
            string[] filters = new string[2];
            filters[0] = type.ToString();
            filters[1] = age.ToString();
            Console.WriteLine("Requesting to filter the participants...");
            SendRequest(ProtoUtils.CreateGetFilteredParticipantsRequest(type, age));
            ContestResponse response = ReadResponse();
            Console.WriteLine("Finished reading the filter response...");
            if (response.Rep == ContestResponse.Types.Reply.Filter)
            {
                return Model2ProtoParticipantMapper(ProtoUtils.GetFilteredPartsDTO(response));
            }
            if (response.Rep == ContestResponse.Types.Reply.Error)
            {
                string err = ProtoUtils.GetError(response);
                throw new ContestException(err);
            }
            return null;
        }

        public int GetParticipantSize()
        {
            throw new NotImplementedException();
        }

        public void Register(string name, int age, contestDomain.enums.Type type1, contestDomain.enums.Type type2)
        {
            Console.WriteLine("Requesting to add a participant...");
            SendRequest(ProtoUtils.CreateRegisterRequest(name, age, type1, type2));
            ContestResponse response = ReadResponse();
            Console.WriteLine("Finished reading the add response...");
            if (response.Rep == ContestResponse.Types.Reply.Error)
            {
                string err = ProtoUtils.GetError(response);
                throw new ContestException(err);
            }
        }

        public List<contestDomain.Task> GetAllTasks()
        {
            throw new NotImplementedException();
        }

        public bool CheckUserExists(string username, string password)
        {
            throw new NotImplementedException();
        }

        public contestDomain.User FindLoggedInUser(string username, string password)
        {
            throw new NotImplementedException();
        }

        public List<contestDomain.Task> GetAllTasksWithMaxEnrolled()
        {
            throw new NotImplementedException();
        }

        public int GetEnrolled(contestDomain.enums.AgeGroup ageGroup, contestDomain.enums.Type type)
        {
            throw new NotImplementedException();
        }

        public List<contestDomain.enums.AgeGroup> GetAllAgeGroups()
        {
            throw new NotImplementedException();
        }

        public List<contestDomain.enums.Type> GetAllTypes()
        {
            throw new NotImplementedException();
        }
    }
}

File: Task Tracker Java/contest_java/contestNetworking/src/main/java/com/example/protobuff/ContestProtoProxy.java
package com.example.protobuff;

import com.example.*;
import com.example.domain.Participant;
import com.example.domain.Task;
import com.example.domain.User;
import com.example.dto.TaskDTO;
import com.example.enums.AgeGroup;
import com.example.enums.Type;

import java.io.*;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class ContestProtoProxy implements IContestService {
    private String host;
    private int port;
    private IContestObserver client;
    private InputStream input;
    private OutputStream output;
    private Socket connection;
    private BlockingQueue<ContestProto.ContestResponse> qresponses;
    private volatile boolean finished;

    public ContestProtoProxy(String host, int port) {
        this.host = host;
        this.port = port;
        qresponses = new LinkedBlockingQueue<ContestProto.ContestResponse>();
    }

    @Override
    public void login(User user, IContestObserver client) throws ContestException, IOException {
        initializeConnection();
        System.out.println("Creating and sending login request...");
        sendRequest(ProtoUtils.createLoginRequest(user));
        System.out.println("Reading the response...");
        ContestProto.ContestResponse response = readResponse();
        System.out.println("Read the response...");
        if (response.getRep() == ContestProto.ContestResponse.Reply.OK) {
            this.client = client;
            return;
        }
        if (response.getRep() == ContestProto.ContestResponse.Reply.Error) {
            String errorText = ProtoUtils.getError(response);
            closeConnection();
            throw new ContestException(errorText);
        }
    }

    @Override
    public void logout(User user, IContestObserver client) throws ContestException {
        System.out.println("Creating and sending logout request...");
        sendRequest(ProtoUtils.createLogoutRequest(user));
        ContestProto.ContestResponse response = readResponse();
        System.out.println("Reading the response...");
        closeConnection();
        if (response.getRep() == ContestProto.ContestResponse.Reply.Error) {
            String errorText = ProtoUtils.getError(response);
            throw new ContestException(errorText);
        }
    }

    private List<Participant> model2protoParticipantMapper(List<ContestProto.ParticipantDTO> parts) {
        List<Participant> res = new ArrayList<>();
        for(var tk: parts){
            Participant p = new Participant(new Random().nextInt(1488 - 500) + 500, tk.getName(), tk.getAge());
            res.add(p);
        }
        return res;
    }

    @Override
    public List<Participant> getFilteredParticipants(Type type, AgeGroup ageGroup) throws ContestException {
        System.out.println("Creating and sending a request to filter participants...");
        sendRequest(ProtoUtils.createGetFilteredParticipantsRequest(type, ageGroup));
        System.out.println("Reading the response...");
        ContestProto.ContestResponse response = readResponse();
        if(response.getRep() == ContestProto.ContestResponse.Reply.FILTER){
            return model2protoParticipantMapper(ProtoUtils.getFilteredPartsDTO(response));
        } else if(response.getRep() == ContestProto.ContestResponse.Reply.Error){
            String error = response.getError() + " eroare";
            throw new ContestException(error);
        }
        return null;
    }

    @Override
    public int getParticipantSize() throws ContestException {
        return 0;
    }

    @Override
    public void register(String name, int age, Type type1, Type type2) throws ContestException, IOException {
        System.out.println("Creating and sending request to add participant... ");
        sendRequest(ProtoUtils.createRegisterRequest(name, age, type1, type2));
        System.out.println("Reading the response...");
        ContestProto.ContestResponse response = readResponse();
        if (response.getRep() == ContestProto.ContestResponse.Reply.Error) {
            String error = response.getError() + " eroare";
            throw new ContestException(error);
        }
    }

    @Override
    public List<Task> getAllTasks() throws ContestException, IOException {
        return null;
    }

    @Override
    public boolean checkUserExists(String username, String password) throws ContestException {
        return false;
    }

    @Override
    public User findLoggedInUser(String username, String password) throws ContestException, IOException {
        return null;
    }

    @Override
    public List<Task> getAllTasksWithMaxEnrolled() throws ContestException {
        return null;
    }

    @Override
    public int getEnrolled(AgeGroup ageGroup, Type type) throws ContestException {
        return 0;
    }

    private List<TaskDTO> model2protoMapper(List<ContestProto.TaskDTO> tasks) {
        List<TaskDTO> res = new ArrayList<>();
        for(var tk: tasks){
            TaskDTO taskDTO = new TaskDTO(tk.getTaskID(), Type.valueOf(tk.getType().toString()), AgeGroup.valueOf(tk.getAge().toString()), tk.getEnrolled());
            res.add(taskDTO);
        }
        return res;
    }

    @Override
    public List<TaskDTO> getTasksEnrolled() throws ContestException {
        System.out.println("Creating and sending tasks enrolled request...");
        sendRequest(ProtoUtils.createGetTasksEnrolledRequest());
        ContestProto.ContestResponse response = readResponse();
        if (response.getRep() == ContestProto.ContestResponse.Reply.GET_ALL_TASKS_EXPERIMENTAL) {
            return model2protoMapper(ProtoUtils.getEnrolledTasks(response));
        } else if (response.getRep() == ContestProto.ContestResponse.Reply.Error) {
            String error = response.getError();
            throw new ContestException(error);
        }
        return null;
    }

    private void initializeConnection() {
        try {
            connection = new Socket(host, port);
            output = connection.getOutputStream();
            input = connection.getInputStream();
            finished = false;
            startReader();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void closeConnection() {
        finished = true;
        try {
            input.close();
            output.close();
            connection.close();
            client = null;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private void startReader() {
        Thread tw = new Thread(new ReaderThread());
        tw.start();
    }

    private void sendRequest(ContestProto.ContestRequest request) throws ContestException {
        try {
            System.out.println("Sending request " + request);
            request.writeDelimitedTo(output);
            output.flush();
        } catch (Exception e) {
            throw new ContestException("Error sending object " + e);
        }
    }

    private ContestProto.ContestResponse readResponse() throws ContestException {
        ContestProto.ContestResponse response;
        try {
            response = qresponses.take();
        } catch (InterruptedException e) {
            throw new ContestException("Error sending object " + e);
        }
        return response;
    }

    private void handleUpdate(ContestProto.ContestResponse response) {
        if (response.getRep() == ContestProto.ContestResponse.Reply.UPDATE) {
            System.out.println("before update");
            try {
                client.updateTaskList();
            } catch (ContestException e) {
                throw new RuntimeException(e);
            }
            System.out.println("after update");
        }
    }

    private boolean isUpdate(ContestProto.ContestResponse response) {
        return response.getRep() == ContestProto.ContestResponse.Reply.UPDATE;
    }

    private class ReaderThread implements Runnable {
        @Override
        public void run() {
            while (!finished) {
                try {
                    ContestProto.ContestResponse response = ContestProto.ContestResponse.parseDelimitedFrom(input);
                    System.out.println("Response received " + response);
                    if (isUpdate(response)) {
                        System.out.println("i'm an update" + response);
                        handleUpdate(response);
                    } else {
                        try {
                            qresponses.put(response);
                        } catch (InterruptedException e) {
                            throw new RuntimeException(e);
                        }
                    }
                } catch (IOException e) {
                    System.out.println("Reading error " + e);
                }
            }
        }
    }
}
